### PR TITLE
Pack of changes to pick from

### PR DESCRIPTION
**Pull request by taublast**
Some changes to browse

1. General:
* added .net8 framework to build along with .net6 version, csproj: 
`<TargetFrameworks>net6.0;net8.0</TargetFrameworks>`

2. HydroComponent + hydro.js:

* added support for custom json serialization settings everywhere, default settings are loop-reference proof now.
```csharp
    public static JsonSerializerSettings JsonSerializerSettings = new()
    {
        Converters = new JsonConverter[] { new Int32Converter() }.ToList(),
        ReferenceLoopHandling = ReferenceLoopHandling.Ignore
    };
```

* Again fixed scroll to anchors after async load, the previous pr-fix was effective in half cases only. inside hydro.js.

```js

 if (push) {
   history.pushState({}, '', url);
 }

   if (selector === findLocationTarget() && !window.location.hash) {
       window.scrollTo(0, 0);
   }
   else {
     
         const element = document.querySelector(window.location.hash);              
          if (element) {
             element.scrollIntoView({
                 behavior: 'smooth', 
                 block: 'start'  
             });
         }

     }
```

* `public void ExecuteJs(string script)` method to invoke js from code behind. Will be executed after the dom is fully rendered. Might consider making it virtual. 

Here hydro.js processes scripts in 2 ways, for static content scripts are present inside the HTML inside custom tags and for dynamic async content they are present inside headers (potential size-limit problem?).  Then scripts are injected like this to be executed by browser:

```js
function injectScript(componentElement, scriptContent) {
  const script = document.createElement('script');
  script.type = 'text/javascript';
  script.innerHTML = `
    (function($component) {
      ${scriptContent}
    })(document.getElementById("${componentElement.id}"));
  `;
  script.setAttribute('data-injected', 'true');
  componentElement.appendChild(script);
}
```

* Execute JS scripts that are included inside components HTML and were not executed in case of ajax-async-loading. Will be executed after the dom is fully rendered.

Present `<script>` elements will never be executed by browser after async loads, so they are removed and at the same location hydro.js creates new script elements.

```js
  function injectScriptAtSamePosition(existingScript) {
    const newScript = document.createElement('script');
    newScript.type = 'text/javascript';
    newScript.innerHTML = existingScript.innerHTML;
    newScript.setAttribute('data-injected', 'true');
    existingScript.parentNode.replaceChild(newScript, existingScript);
}
```

* Added native events that are invoked by hydro.js to easily debug and plug-in your JS code:

example of subscribing to:

```js
document.addEventListener('HydroLocation', function (event) {
            
        const { url, headers, selector, push, payload } = event.detail;
            
        console.log("HydroLocation:", event.detail);

        initControls();

    });

document.addEventListener('HydroComponentUpdate', function (event) {

    const { componentId, name, url, type } = event.detail;
            
    console.log("HydroComponentUpdate: ", event.detail);

    //if (name == "KbDisplayComponent") {
    //    scrollToTop();
    //}

});

document.addEventListener('HydroRedirect', function (event) {

    const { url } = event.detail;
            
    console.log("HydroRedirect: ", event.detail);

    mainMenu();
});

document.addEventListener('HydroComponentInit', function (event) {

    const { component } = event.detail;
            
    console.log("HydroComponentInit: ", event.detail);

});

```

* `With ExecuteJs` we can now do the following. This is NOT in the PR but IMHO should take it's place inside the hydro core.

Example: you have finished editing something and you want to update a specific component to reflect the changes. From component A just call `ExecuteJs("updateComponent('ComponentB')");`

pick up inside JS:

```js
//hacky update for ANY component. could make a clean method.
async function updateComponent(findName) {
    let element = document.querySelector('[hydro-name="'+findName+'"]');
    await window.Hydro.hydroAction(element, window.Hydro.findComponent(element), { name: "Mount"});
}

```
If this goes into core we then could add another native event `HydroNeedUpdate`, could be subscribed to as follows:

```js
document.addEventListener('HydroNeedUpdate', function (event) {
    const { name } = event.detail;            
    console.log("HydroNeedUpdate: ", event.detail);

//an update was ordered for a specific component
  
});

```

* Added some blazor-friendly virtual methods to favor porting blazor code more easily to hydro. 

`virtual Task OnAfterRenderAsync(bool firstRender)`
`protected virtual Task OnParametersSetAsync()`
`protected virtual bool ShouldRender()`

* added ICookieManager for fluent cookies management

* added customization to Location that would be replaced after async load, ability to have _Layout=null and for a faster async loading of updated components

Example: 
1. In your _Laout replace `@RenderBody()` with `@Html.HydroBody(RenderBody())`. Why so that hydro can automatically recognize this location and insert all async loaded HTML here instead of replacing the whole `<body>`.
2. Inside your _ViewStart.cshtml replace `Layout = "_Layout";` with
```csharp
@using Hydro
@{
    Layout = Context.IsHydroBoosted() ? null : "_Layout";
}
```
Now you are all set for faster component rendering avoiding to render the layout for ajax async calls.
At the same time notice an additional method `public void Location(string url, string target, object payload = null)` where `target` can be anything, "#mybody", ".wrapper" etc.

* Ability to POST a `<form>` and validate/read inside the action:

```csharp
        public async Task Submit()
        {
            //can read form fields manually
            var email = this.Request.Form["Data.Email"];

           //method to auto-read, returns value of ModelState.IsValid
           var isValid = await BindFormAndValidateAsync(Data, (model, modelState) =>
           {
               //optional code to transform received fields before they go to validation
               if (model.Name != null)
                   model.Name = model.Name.ToPhraseCase().Trim();
        
               //can check email is unique
               if (!EmailValidator.IsValidEmail(model.Email))
               {
                   modelState.AddModelError("Email", "Email taken");
               }
           });

            if (isValid)
            {
                Error=""; //displayed in UI

                try
                {
                    //can read posted files too
                    if (Request.Form.Files.Count > 0)
                    {
                        var savedAs = await SaveAttachmentFile("avatar");
                        if (!string.IsNullOrEmpty(savedAs))
                        {
                            Data.AvatarPath = savedAs;
                        }
                    }
                    
                     ...
                  
                }
                catch (Exception e)
                {
                    Console.WriteLine(e);
                    this.Error = e.Message;
                }
            }
            else
            {
                var errors = ModelState.Values
                .SelectMany(v => v.Errors)
                .Select(e => e.ErrorMessage)
                .ToList();

                Error = errors.ToTags(", ");
            }
        }
```

some points inside hydro.js:

```js
//inside enqueueHydroPromise
    // NEW post form from eventData
    if (eventData instanceof HTMLFormElement) {
    new FormData(eventData).forEach((value, key) => {
        requestForm.append(key, value);
    });
```

```js
  Alpine.data('hydro', () => {

      ...

      async invoke(e, action) {
        if (["click", "submit"].includes(e.type) && ['A', 'BUTTON', 'FORM'].includes(this.$el.tagName)) {
          e.preventDefault();
        }
        await window.Hydro.hydroAction(this.$el, this.$component, action);
      },
      async postForm(e, action) {  // this will be called by code from HydroOnTagHelper->GetJsPostFormExpression
        if (["click", "submit"].includes(e.type) && ['A', 'BUTTON', 'FORM'].includes(this.$el.tagName)) {
          e.preventDefault();
        }
        await window.Hydro.hydroAction(this.$el, this.$component, action, e.target);
      },
```

### CODE REVIEW

**Comment by kjeske** [comment]
Thanks for all the descriptions! :) I will keep this PR as a reference and create a separate one (or more) with changes related to specific features, while keeping your attribution on the release.

File: src/HydroComponent.cs [hunk 1]
```diff
@@ -382,8 +451,10 @@ private async Task<string> RenderOnlineRootComponent(IPersistentState persistent
 
         PopulateDispatchers();
 
+        PopulateClientScripts();
+
         return !_skipOutput
-            ? await GenerateComponentHtml(componentId, persistentState)
+            ? await GenerateComponentHtml(componentId, persistentState, false)
```
Comment (kjeske) [inline src/HydroComponent.cs:457]: Why not execute scripts when rendering root component on Hydro request (isStatic = false), but execute on nested components (isStatic = true)?

**Comment by taublast** [comment]
Hi! As mentioned before we have 2 different ways for rendering components
and thus executing scripts: 1 static prepared content for full page load
and 2 ajax-loaded dynamic updates. As I recall in this example our root
component carries its children nested components that would have the
"static pre-rendered" look (way 2), while the parent itself is loaded with
async update (way 1). So scripts for children here should be
treated-processed the same way as is they were inside the first static page
load while the parent is going under a different procedure. Hmm might not
sound very clear, but still.. :)

**Comment by taublast** [comment]
An important point on "how to use" the form submit to pass all field with form:

A - Usual Hydro way: put `hydro-on:click="@(() => Model.Submit())"` into the submit button.

B - To pass all form fields: 
1. the submit button doesn't need to call hydro action, it must contain a `submit` attribute instead.
2. the <form> must contain `hydro-on:submit="@(() => Model.Submit())"`

**Comment by kjeske** [comment]
@taublast I think it would be nice to keep calling the actions in an explicit way, so on form:
```html
<form hydro-on:submit="@(() => Model.Submit())">
```
or on button
```html
<button type="submit" hydro-on:click="@(() => Model.Submit())">
```

Many might not want at this stage to pass any fields, since all of them went already using `hydro-bind`. That's why sending the form fields on submit would have to be a conscious decision. I think I mentioned before some ideas. I would go for putting an modifier `bind` after submit which could be read in hydro JS lib.

```razor
<form hydro-on:submit.bind="@(() => Model.Submit())">

</form>
```

`bind` modifier would be applicable only for `submit` event.

**Pull request closed without merging**